File: java/src/test/java/project_euler/example/problem/P004.java
package project_euler.example.problem;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * Problem 4 「最大の回文積」
 *
 * <pre>
 * 左右どちらから読んでも同じ値になる数を回文数という. 2桁の数の積で表される回文数のうち, 最大のものは 9009 = 91 × 99 である.
 *
 * では, 3桁の数の積で表される回文数の最大値を求めよ.
 * </pre>
 *
 * @see <a href="https://odz.sakura.ne.jp/projecteuler/?Problem+4" target="_blank">問題リンク</a>
 */
public class P004 {

  private static long answer(long min, long max) {
    var result = 0L;
    for (long i = max; min <= i; i--) {
      for (long j = max; min <= j; j--) {
        var calc = i * j;
        var left = String.valueOf(calc);
        var right = new StringBuilder(left).reverse().toString();
        if (left.equals(right) && result < calc) {
          result = calc;
        }
      }
    }
    return result;
  }

  @Test
  void テスト1() {
    assertEquals(9009, answer(10, 99));
  }

  @Test
  void テスト2() {
    assertEquals(906609, answer(100, 999));
  }
}

File: csharp/project-euler/Test/P004.cs
using System;
using System.Linq;
using NUnit.Framework;

namespace project_euler.Test
{
  /// <summary>
  /// <a href="http://odz.sakura.ne.jp/projecteuler/?Problem+4">Problem 4 「最大の回文積」 </a>
  /// <remarks>
  /// 左右どちらから読んでも同じ値になる数を回文数という. 2桁の数の積で表される回文数のうち, 最大のものは 9009 = 91 × 99 である.
  ///
  /// では, 3桁の数の積で表される回文数の最大値を求めよ.
  /// </remarks>
  /// </summary>
  public class P004
  {
    private static ulong Run(ulong min, ulong max)
    {
      var result = 0ul;
      for (var i = max; min <= i; i--)
      {
        for (var j = max; min <= j; j--)
        {
          var calc = i * j;
          var left = calc.ToString();
          var right = string.Concat(left.Reverse());
          if (left == right && result < calc) {
            result = calc;
          }
        }
      }
      return result;
    }

    [Test]
    public void テスト1()
    {
      Assert.AreEqual(9009, Run(10, 99));
    }

    [Test]
    public void テスト2()
    {
      Assert.AreEqual(906609, Run(100, 999));
    }
  }
}
